pandas/core/arrays/boolean.py

```diff
@@ -0,0 +1,725 @@
+            with warnings.catch_warnings():
+                warnings.filterwarnings("ignore", "elementwise", FutureWarning)
+                with np.errstate(all="ignore"):
+                    result = op(self._data, other)
+
+            # nans propagate
+            if mask is None:
+                mask = self._mask
+            else:
+                mask = self._mask | mask
+
+            return BooleanArray(result, mask)
+
+        name = "__{name}__".format(name=op.__name__)
+        return set_function_name(logical_method, name, cls)
+
+    @classmethod
+    def _create_comparison_method(cls, op):
+        op_name = op.__name__
+
+        def cmp_method(self, other):
+
+            if isinstance(other, (ABCDataFrame, ABCSeries, ABCIndexClass)):
+                # Rely on pandas to unbox and dispatch to us.
+                return NotImplemented
+
+            other = lib.item_from_zerodim(other)
+            mask = None
+
+            if isinstance(other, BooleanArray):
+                other, mask = other._data, other._mask
+
+            elif is_list_like(other):
+                other = np.asarray(other)
+                if other.ndim > 1:
+                    raise NotImplementedError(
+                        "can only perform ops with 1-d structures"
+                    )
+                if len(self) != len(other):
+                    raise ValueError("Lengths must match to compare")
+
+            # numpy will show a DeprecationWarning on invalid elementwise
+            # comparisons, this will raise in the future
+            with warnings.catch_warnings():
+                warnings.filterwarnings("ignore", "elementwise", FutureWarning)
+                with np.errstate(all="ignore"):
+                    result = op(self._data, other)
+
+            # nans propagate
+            if mask is None:
+                mask = self._mask
+            else:
+                mask = self._mask | mask
+
+            result[mask] = op_name == "ne"
+            return BooleanArray(result, np.zeros(len(result), dtype=bool), copy=False)
+
+        name = "__{name}__".format(name=op.__name__)
+        return set_function_name(cmp_method, name, cls)
+
+    def _reduce(self, name, skipna=True, **kwargs):
+        data = self._data
+        mask = self._mask
+
+        # coerce to a nan-aware float if needed
+        if mask.any():
+            data = self._data.astype("float64")
+            data[mask] = self._na_value
+
+        op = getattr(nanops, "nan" + name)
+        result = op(data, axis=0, skipna=skipna, mask=mask)
+
+        # if we have a boolean op, don't coerce
+        if name in ["any", "all"]:
+            pass
+
+        # if we have numeric op that would result in an int, coerce to int if possible
+        elif name in ["sum", "min", "max", "prod"] and notna(result):
+            int_result = int(result)
+            if int_result == result:
+                result = int_result
+
```

Comment:
what about other names? do we raise?

Reply:
> what about other names? do we raise?

No, they work, but the result of the numpy operation already has the correct dtype (eg std, var, .. results in a float), so they don't need to be post-processed.

(but just realizing that min/max should return boolean not an int. For sum and prod the int is correct)